terra-base-jupyter/Dockerfile

```diff
@@ -0,0 +1,199 @@
+    && locale-gen \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+# Install Node >18 (needed for jupyterlab)
+RUN apt-get update && apt-get install -yq --no-install-recommends 
+RUN mkdir -p /etc/apt/keyrings
+RUN curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg
+
+RUN echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_$NODE_MAJOR.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list
+RUN dpkg --remove --force-remove-reinstreq libnode-dev
+RUN apt-get update && apt-get install -f -yq nodejs
+
+#############
+# Users Setup
+#############
+# Create the welder user
+# The welder uid is consistent with the Welder docker definition here:
+#  https://github.com/DataBiosphere/welder/blob/master/project/Settings.scala
+# Adding welder-user to the Jupyter container isn't strictly required, but it makes welder-added
+# files display nicer when viewed in a terminal.
+ENV WELDER_USER welder-user
+ENV WELDER_UID 1001
+RUN useradd -m -N -u $WELDER_UID $WELDER_USER
+
+# Create the jupyter user
+ENV USER jupyter
+ENV USER_HOME /home/$USER
+RUN useradd -m -d $USER_HOME -N -g users $USER
+# We want to grant the jupyter user sudo permissions
+# without password so they can install the necessary packages that they 
+# want to use on the docker container
+RUN echo "$USER ALL=(ALL) NOPASSWD: ALL" > /etc/sudoers.d/$USER \
+    && chmod 0440 /etc/sudoers.d/$USER
+
+#####################################
+# Install Python 3.10 via Miniconda
+#####################################
+## Note: CONDA should NOT be used by terra devs to manage dependencies (see the use of poetry below instead),
+## but is a widely used tool to manage python environments in a runtime and we should provide it to users
+## We want to store the user conda environments in a directory
+## that will be in the persistent disk
+## Attention: If you change the Conda home location, please update conda_init.txt accordingly
+ENV CONDA_ENV_NAME base-python3.10
+ENV CONDA_ENV_HOME $USER_HOME/.envs/$CONDA_ENV_NAME
+RUN curl -so $HOME/miniconda.sh https://repo.anaconda.com/miniconda/Miniconda3-py310_23.5.1-0-Linux-x86_64.sh \
+    && chmod +x $HOME/miniconda.sh \
+    && $HOME/miniconda.sh -b -p $CONDA_ENV_HOME \
+    && rm $HOME/miniconda.sh
+ENV PATH "${PATH}:${CONDA_ENV_HOME}/bin"
+
+ENV BASE_PYTHON_PATH $CONDA_ENV_HOME/bin/python
+ENV PYTHONDONTWRITEBYTECODE=true
+
+###################################################
+# Set up the user to use the conda base environment
+###################################################
+## The user should have full access to the conda base environment, and can use it directly, or
+## create new conda environments on top of it. The important part is that jupyter IS NOT installed
+## in the base environment to provide isolation between the user environment, and the jupyter server
+## to avoid cross-contamination
+COPY conda-environment.yml .
```

Comment:
If I understand correctly the goal here, maybe using conda-pack can help further reduce the image size.
When we build dockers that uses conda to manage envs, we use conda-pack combined with multi-stage builds to reduce image sizes.
Example here:
https://github.com/broadinstitute/long-read-pipelines/blob/4b50b3857d33fd195461e5eb5c8a83d7fe6dda27/docker/lr-papermill-base/Dockerfile#L8

Reply:
To add to this, it'd be good to have an understanding how large each layer is, before optimizing the sizes further.

Reply:
That is a good call out. I will add a detail of how big each layer is in the Readme. Regarding conda-pack, I definitely get why you are using it, but since the plan is to have the majority of Terra user use this base image, I was thinking about setting up the base conda environment directly.
The use case of building a custom image on top of it is a bit of an edge case, and so the base image might not be perfectly curated for your need. Would this be a problem?

Reply:
No, this would not be a problem, just an optimization (which could be evil as it might be pre-mature). OTH, our experience with `conda-pack` is that it reduces the image size further. 
This is where we borrowed the lesson. https://pythonspeed.com/articles/conda-docker-image-size/

I can give it a try once the other components are steady, actually since I'm the one mostly interested in it.